src/compiler/checker.ts

```diff
@@ -14204,22 +14204,27 @@ namespace ts {
                         getFlowTypeOfReference(accessExpression, propType) :
                         propType;
                 }
-                if (everyType(objectType, isTupleType) && isNumericLiteralName(propName) && +propName >= 0) {
-                    if (accessNode && everyType(objectType, t => !(<TupleTypeReference>t).target.hasRestElement) && !(accessFlags & AccessFlags.NoTupleBoundsCheck)) {
-                        const indexNode = getIndexNodeForAccessExpression(accessNode);
-                        if (isTupleType(objectType)) {
-                            error(indexNode, Diagnostics.Tuple_type_0_of_length_1_has_no_element_at_index_2,
-                                typeToString(objectType), getTypeReferenceArity(objectType), unescapeLeadingUnderscores(propName));
-                        }
-                        else {
-                            error(indexNode, Diagnostics.Property_0_does_not_exist_on_type_1, unescapeLeadingUnderscores(propName), typeToString(objectType));
+                if (isNumericLiteralName(propName) && +propName >= 0) {
+                    if (everyType(objectType, isTupleType)) {
+                        if (accessNode && everyType(objectType, t => !(<TupleTypeReference>t).target.hasRestElement) && !(accessFlags & AccessFlags.NoTupleBoundsCheck)) {
+                            const indexNode = getIndexNodeForAccessExpression(accessNode);
+                            if (isTupleType(objectType)) {
+                                error(indexNode, Diagnostics.Tuple_type_0_of_length_1_has_no_element_at_index_2,
+                                    typeToString(objectType), getTypeReferenceArity(objectType), unescapeLeadingUnderscores(propName));
+                            }
+                            else {
+                                error(indexNode, Diagnostics.Property_0_does_not_exist_on_type_1, unescapeLeadingUnderscores(propName), typeToString(objectType));
+                            }
                         }
+                        errorIfWritingToReadonlyIndex(getIndexInfoOfType(objectType, IndexKind.Number));
+                        return mapType(objectType, t => {
+                            const restType = getRestTypeOfTupleType(<TupleTypeReference>t) || undefinedType;
+                            return noUncheckedIndexedAccessCandidate ? getUnionType([restType, undefinedType]) : restType;
+                        });
+                    }
+                    else if (someType(objectType, isTupleType)) {

```

Comment:
This returns `undefined` for something like `{[idx: number]: number} | [string]`, which is incorrect (the object type has a numeric indexer, after all) - at worst, you'd want something like `number | undefined` for an index of `1` or more, and not `undefined`. Moreover, an intersection, like `{[idx: number]: number} & [string]` should likewise produce `number` for the result of indexing `1`, and not `undefined`.